### PR TITLE
LPAL-21 Zend to Laminas upgrade

### PR DESCRIPTION
**Pull request by nickdavis2001**
## Purpose
_LPAL-21 Upgrade zend to laminas_

## Approach

- Ran the automatic migration script from Zend to Laminas. 
- In composer.json for the various components, Changed php to 7.4 upgraded a handful of components missed by the migrator, brought versions more up to date and closer to Use's version, added the zend-laminas-bridge component which maps namespaces. 
- ZfcRbac was renamed to LmcRbacMvc, so updated many files by hand to reflect this. 
- Twig components were renamed. 
- Had to isolate some Twig code which didn't play well wth Mockery due to a class declared as final
- ZendPdf is left at original version because this is the one that currently works with this setup.

_

## Learning

_Any tips and tricks, blog posts or tools which helped you. Plus anything notable you've discovered about the LPA service_

## Checklist

* [x] I have performed a self-review of my own code
* [ ] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
* [ ] I have added tests to prove my work
* [ ] The product team have tested these changes

### CODE REVIEW

File: service-front/config/application.config.php [hunk 1]
```diff
@@ -2,19 +2,19 @@
...
+        'Laminas\InputFilter',
+        'Laminas\Filter',
+        'Laminas\Hydrator',
+        'Laminas\Cache',
+        'Laminas\Session',
+        'Laminas\Router',
+        'Laminas\Validator',
         'Application',
         //ZfcTwig MUST be the last module!
         'ZfcTwig'
```
Comment (williamfalconeruk) [inline service-front/config/application.config.php:20]: is zfcTwig right here? is there a laminas replacement?

File: service-front/module/Application/src/Form/Validator/Csrf.php [hunk 2]
```diff
@@ -2,9 +2,9 @@
 
 namespace Application\Form\Validator;
 
-use Zend\Math\Rand;
-use Zend\Session\Container;
-use Zend\Validator\Csrf as ZFCsrfValidator;
+use Laminas\Math\Rand;
+use Laminas\Session\Container;
+use Laminas\Validator\Csrf as ZFCsrfValidator;
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/Form/Validator/Csrf.php:7]: maybe rename this alias

File: service-front/module/Application/src/Form/View/Helper/FormMultiCheckbox.php [hunk 3]
```diff
@@ -2,9 +2,9 @@
 
 namespace Application\Form\View\Helper;
 
-use Zend\Form\Element\MultiCheckbox;
-use Zend\Form\LabelAwareInterface;
-use Zend\Form\View\Helper\FormMultiCheckbox as ZFFormMultiCheckbox;
+use Laminas\Form\Element\MultiCheckbox;
+use Laminas\Form\LabelAwareInterface;
+use Laminas\Form\View\Helper\FormMultiCheckbox as ZFFormMultiCheckbox;
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/Form/View/Helper/FormMultiCheckbox.php:7]: maybe rename this alias

File: service-front/module/Application/src/Form/View/Helper/FormRadio.php [hunk 4]
```diff
@@ -2,10 +2,10 @@
...
 namespace Application\Form\View\Helper;
 
-use Zend\Form\Element\MultiCheckbox;
-use Zend\Form\Element\Radio;
-use Zend\Form\LabelAwareInterface;
-use Zend\Form\View\Helper\FormRadio as ZFFormRadioHelper;
+use Laminas\Form\Element\MultiCheckbox;
+use Laminas\Form\Element\Radio;
+use Laminas\Form\LabelAwareInterface;
+use Laminas\Form\View\Helper\FormRadio as ZFFormRadioHelper;
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/Form/View/Helper/FormRadio.php:8]: maybe rename this alias

File: service-front/module/Application/src/Form/Validator/EmailAddress.php [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 namespace Application\Form\Validator;
 
-use Zend\Validator\EmailAddress as ZfEmailAddressValidator;
+use Laminas\Validator\EmailAddress as ZfEmailAddressValidator;
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/Form/Validator/EmailAddress.php:5]: maybe rename this alias

File: service-front/module/Application/src/Model/Service/Authentication/Adapter/AdapterInterface.php [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 namespace Application\Model\Service\Authentication\Adapter;
 
-use Zend\Authentication\Adapter\AdapterInterface as ZFAdapterInterface;
+use Laminas\Authentication\Adapter\AdapterInterface as ZFAdapterInterface;
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/Model/Service/Authentication/Adapter/AdapterInterface.php:5]: maybe rename this alias

File: service-front/module/Application/src/Model/Service/Mail/Message.php [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 namespace Application\Model\Service\Mail;
 
-use Zend\Mail\Message as ZFMessage;
+use Laminas\Mail\Message as ZFMessage;
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/Model/Service/Mail/Message.php:5]: maybe rename this alias

File: service-front/module/Application/src/Model/Service/Mail/Transport/MailTransport.php [hunk 8]
```diff
@@ -6,16 +6,16 @@
...
-use Zend\Mail\Exception\InvalidArgumentException;
-use Zend\Mail\Header\GenericHeader;
-use Zend\Mail\Message as ZFMessage;
-use Zend\Mail\Transport\Exception\InvalidArgumentException as TransportInvalidArgumentException;
-use Zend\Mail\Transport\TransportInterface;
-use Zend\Mime\Message as MimeMessage;
+use Laminas\Mime;
+use Laminas\Mail\Exception\InvalidArgumentException;
+use Laminas\Mail\Header\GenericHeader;
+use Laminas\Mail\Message as ZFMessage;
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/Model/Service/Mail/Transport/MailTransport.php:12]: maybe rename this alias

File: service-front/module/Application/src/Model/Service/Session/SessionManager.php [hunk 9]
```diff
@@ -1,8 +1,8 @@
 <?php
 namespace Application\Model\Service\Session;
 
-use Zend\Session\Container;
-use Zend\Session\SessionManager as ZFSessionManager;
+use Laminas\Session\Container;
+use Laminas\Session\SessionManager as ZFSessionManager;
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/Model/Service/Session/SessionManager.php:5]: maybe rename this alias

File: service-front/module/Application/src/View/Helper/LocalViewRenderer.php [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * @param TwigEnvironment $viewRenderer
+     */
+    public function __construct(TwigEnvironment $viewRenderer)
+    {
+        $this->viewRenderer = $viewRenderer;
+    }
+
+    public function LoadTemplate(string $templateName)
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/View/Helper/LocalViewRenderer.php:22]: @inheritdoc needed here

File: service-front/module/Application/src/View/Helper/LocalViewRenderer.php [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    {
+         return $this->viewRenderer->load($templateName)->unwrap();
+    }
+}
+
+interface RendererInterface
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/View/Helper/LocalViewRenderer.php:28]: split into separate file possibly - add autodocs here and use @inheritdoc on the class implementing this interface.

File: service-front/module/Application/src/Form/Validator/Csrf.php (hunk 2 above)
Comment (nickdavis2001) [reply service-front/module/Application/src/Form/Validator/Csrf.php:7]: OK. Replacing ZF with Laminas

File: service-front/module/Application/src/Form/View/Helper/FormMultiCheckbox.php (hunk 3 above)
Comment (nickdavis2001) [reply service-front/module/Application/src/Form/View/Helper/FormMultiCheckbox.php:7]: OK. Replacing ZF with Laminas

File: service-front/module/Application/src/Form/View/Helper/FormRadio.php (hunk 4 above)
Comment (nickdavis2001) [reply service-front/module/Application/src/Form/View/Helper/FormRadio.php:8]: OK. Replacing ZF with Laminas

File: service-front/module/Application/src/Form/Validator/EmailAddress.php (hunk 5 above)
Comment (nickdavis2001) [reply service-front/module/Application/src/Form/Validator/EmailAddress.php:5]: OK. Replacing ZF with Laminas

File: service-front/module/Application/src/Model/Service/Authentication/Adapter/AdapterInterface.php (hunk 6 above)
Comment (nickdavis2001) [reply service-front/module/Application/src/Model/Service/Authentication/Adapter/AdapterInterface.php:5]: OK. Replacing ZF with Laminas

File: service-front/module/Application/src/Model/Service/Mail/Message.php (hunk 7 above)
Comment (nickdavis2001) [reply service-front/module/Application/src/Model/Service/Mail/Message.php:5]: OK. Replacing ZF with Laminas

File: service-front/module/Application/src/Model/Service/Mail/Transport/MailTransport.php (hunk 8 above)
Comment (nickdavis2001) [reply service-front/module/Application/src/Model/Service/Mail/Transport/MailTransport.php:12]: OK. Replacing ZF with Laminas

File: service-front/module/Application/src/Model/Service/Session/SessionManager.php (hunk 9 above)
Comment (nickdavis2001) [reply service-front/module/Application/src/Model/Service/Session/SessionManager.php:5]: OK. Replacing ZF with Laminas

File: service-front/config/application.config.php (hunk 1 above)
Comment (nickdavis2001) [reply service-front/config/application.config.php:20]: Good Q but in fact the Twig plugin for Laminas is still called ZfcTwig.  In an ideal world I guess they would rename that too. Just for reassurance, Rich McHale told me Sirius after upgrade is using the exact same ZfcTwig version as we are moving to here.

File: service-front/module/Application/src/View/Helper/RendererInterface.php [hunk 12]
```diff
@@ -0,0 +1,10 @@
+<?php
+// @inheritdoc
```
Comment (williamfalconeruk) [inline service-front/module/Application/src/View/Helper/RendererInterface.php:2]: remove this comment, and replace with autodocs. see https://www.jetbrains.com/help/phpstorm/phpdoc-comments.html if this helps.

Comment (williamfalconeruk) [reply service-front/module/Application/src/View/Helper/RendererInterface.php:2]: also see https://manual.phpdoc.org/HTMLSmartyConverter/HandS/phpDocumentor/tutorial_tags.inlineinheritdoc.pkg.html

Comment (nickdavis2001) [reply service-front/module/Application/src/View/Helper/RendererInterface.php]: Done

File: service-front/module/Application/src/View/Helper/LocalViewRenderer.php (hunk 11 above)
Comment (nickdavis2001) [reply service-front/module/Application/src/View/Helper/LocalViewRenderer.php]: Done

File: service-front/module/Application/src/View/Helper/LocalViewRenderer.php (hunk 10 above)
Comment (nickdavis2001) [reply service-front/module/Application/src/View/Helper/LocalViewRenderer.php:20]: Done

**Review by williamfalconeruk — APPROVED** [review]

**Comment by williamfalconeruk** [comment]
🏆  🎖️ 🥇

**Pull request merged**